File: webmcp/src/agent.ts
/**
 * Agent-intuitive VCP Agent Runtime Profile contracts.
 *
 * This module is a source candidate. It performs no implicit network access and
 * carries no authority. Hosts inject controlled and accretive services.
 */

export type AgentProfile =
	| 'observe@0.1.0'
	| 'controlled@0.1.0'
	| 'accretive@0.1.0';

export interface AgentRuntimeProfileOffer {
	kind: 'agent_runtime_profile_offer';
	version: '0.1.0';
	required: readonly AgentProfile[];
	optional: readonly AgentProfile[];
}

export interface AgentRuntimeProfileAcknowledgement {
	kind: 'agent_runtime_profile_ack';
	version: '0.1.0';
	selected: readonly AgentProfile[];
	unsupported_optional: readonly AgentProfile[];
	bootstrap_ref: string;
	capability_catalog_digest: string;
	principal_session_ref: string;
	event_binding: 'cursor' | 'stream' | 'poll';
	expires_at: string;
}

export type AgentResultStatus =
	| 'ready'
	| 'degraded'
	| 'awaiting_review'
	| 'blocked'
	| 'unavailable'
	| 'stale'
	| 'conflicting'
	| 'budget_exhausted'
	| 'indeterminate'
	| 'failed';

export type EffectClass =
	| 'pure_local'
	| 'state_read'
	| 'sensitive_egress'
	| 'reversible_write'
	| 'communication'
	| 'financial'
	| 'legal'
	| 'physical';

export interface ResourceBudget {
	wall_time_ms: number | null;
	tokens: number | null;
	external_calls: number | null;
	money_minor: number | null;
	human_interruptions: number | null;
	reserve_fraction: number;
	model_calls: number | null;
	local_compute_ms: number | null;
	bytes: number | null;
	sensitive_egress_bytes: number | null;
	privacy_units: number | null;
	risk_units: number | null;
	welfare_load_units: number | null;
}

export interface SituationView {
	kind: 'situation_view';
	version: '0.1.0';
	situation_id: string;
	goal: string;
	principal_ref: string;
	known_claim_refs: readonly string[];
	unknowns: readonly string[];
	conflict_refs: readonly string[];
	normative_context_ref: string;
	authority_refs: readonly string[];
	budget: ResourceBudget;
	active_work_refs: readonly string[];
	control_operations: readonly string[];
	affordance_refs: readonly string[];
	omissions: readonly unknown[];
	as_of: string;
	cursor: string;
	dependency_digest: string;
	digest: string;
}

export interface ActionIntent {
	kind: 'action_intent';
	version: '0.1.0';
	intent_id: string;
	run_ref: string;
	step_ref: string;
	affordance_ref: string;
	arguments_digest: string;
	destination: string;
	context_digest: string;
	policy_digest: string;
	descriptor_digest: string;
	requested_at: string;
	digest: string;
	schema_digest: string;
	effect_class: EffectClass;
	situation_digest: string;
	expected_postconditions: readonly string[];
	resource_ceiling: ResourceBudget;
	idempotency_scope: string;
	requested_authority: string;
}

export interface ExecutionReceipt {
	kind: 'execution_receipt';
	version: '0.1.0';
	receipt_id: string;
	attempt_ref: string;
	effect_status:
		| 'none'
		| 'accepted'
		| 'observed'
		| 'failed'
		| 'possible'
		| 'indeterminate'
		| 'compensated';
	provider_ref: string | null;
	evidence_refs: readonly string[];
	observed_at: string;
	reconcile_ref: string | null;
	digest: string;
}

export interface AccretionCandidate {
	kind: 'accretion_candidate';
	version: '0.1.0';
	candidate_id: string;
	candidate_kind: string;
	content: unknown;
	scope: readonly string[];
	provenance_refs: readonly string[];
	validation_status: string;
	review_required: boolean;
	expires_at?: string | null;
	digest: string;
	source_run_ref: string;
	supporting_evidence_refs: readonly string[];
	contradicting_evidence_refs: readonly string[];
	sensitivity: string;
	confidence: number;
	invalidation_triggers: readonly string[];
	revalidation: string;
	promotion_policy: string;
	expected_utility: number;
	rollback: string;
	quarantine_status: string;
	dependency_digest: string;
}

export interface PromotionRecord {
	kind: 'promotion_record';
	version: '0.1.0';
	promotion_id: string;
	candidate_ref: string;
	promoted_asset_ref: string;
	authority_ref: string;
	decision_ref: string;
	promoted_at: string;
	expires_at?: string | null;
	revocation_ref: string;
	digest: string;
	evidence_refs: readonly string[];
	validation_results: readonly string[];
	scope: readonly string[];
	promoted_content_digest: string;
	dependency_digest: string;
}

export type AgentArtifact =
	| SituationView
	| ActionIntent
	| ExecutionReceipt
	| AccretionCandidate
	| PromotionRecord;

export type AgentRuntimeDocument =
	| AgentRuntimeProfileOffer
	| AgentRuntimeProfileAcknowledgement
	| AgentArtifact;

export interface AgentFailure {
	code: string;
	category: string;
	summary: string;
	retry: string;
}

export interface AgentResult<T> {
	status: AgentResultStatus;
	value: T | null;
	failure: AgentFailure | null;
	warnings: readonly string[];
	evidence_refs: readonly string[];
}

const ARTIFACT_KEYS: Record<AgentArtifact['kind'], ReadonlySet<string>> = {
	situation_view: new Set([
		'kind', 'version', 'situation_id', 'goal', 'principal_ref',
		'known_claim_refs', 'unknowns', 'conflict_refs', 'normative_context_ref',
		'authority_refs', 'budget', 'active_work_refs', 'control_operations',
		'affordance_refs', 'omissions', 'as_of', 'cursor', 'dependency_digest', 'digest'
	]),
	action_intent: new Set([
		'kind', 'version', 'intent_id', 'run_ref', 'step_ref', 'affordance_ref',
		'arguments_digest', 'destination', 'context_digest', 'policy_digest',
		'descriptor_digest', 'requested_at', 'digest', 'schema_digest', 'effect_class',
		'situation_digest', 'expected_postconditions', 'resource_ceiling',
		'idempotency_scope', 'requested_authority'
	]),
	execution_receipt: new Set([
		'kind', 'version', 'receipt_id', 'attempt_ref', 'effect_status', 'provider_ref',
		'evidence_refs', 'observed_at', 'reconcile_ref', 'digest'
	]),
	accretion_candidate: new Set([
		'kind', 'version', 'candidate_id', 'candidate_kind', 'content', 'scope',
		'provenance_refs', 'validation_status', 'review_required', 'expires_at',
		'digest', 'source_run_ref', 'supporting_evidence_refs',
		'contradicting_evidence_refs', 'sensitivity', 'confidence',
		'invalidation_triggers', 'revalidation', 'promotion_policy',
		'expected_utility', 'rollback', 'quarantine_status', 'dependency_digest'
	]),
	promotion_record: new Set([
		'kind', 'version', 'promotion_id', 'candidate_ref', 'promoted_asset_ref',
		'authority_ref', 'decision_ref', 'promoted_at', 'expires_at',
		'revocation_ref', 'digest', 'evidence_refs', 'validation_results', 'scope',
		'promoted_content_digest', 'dependency_digest'
	])
};

const NEGOTIATION_KEYS: Record<
	AgentRuntimeProfileOffer['kind'] | AgentRuntimeProfileAcknowledgement['kind'],
	ReadonlySet<string>
> = {
	agent_runtime_profile_offer: new Set(['kind', 'version', 'required', 'optional']),
	agent_runtime_profile_ack: new Set([
		'kind', 'version', 'selected', 'unsupported_optional', 'bootstrap_ref',
		'capability_catalog_digest', 'principal_session_ref', 'event_binding', 'expires_at'
	])
};

const PROFILE_IDS = new Set<AgentProfile>([
	'observe@0.1.0',
	'controlled@0.1.0',
	'accretive@0.1.0'
]);

function objectValue(input: unknown): Record<string, unknown> {
	if (typeof input !== 'object' || input === null || Array.isArray(input)) {
		throw new TypeError('agent artifact must be an object');
	}
	return input as Record<string, unknown>;
}

/** Parse a portable artifact without accepting undeclared authority fields. */
export function parseAgentArtifact(input: unknown): AgentArtifact {
	const document = parseAgentRuntimeDocument(input);
	if (
		document.kind === 'agent_runtime_profile_offer' ||
		document.kind === 'agent_runtime_profile_ack'
	) {
		throw new TypeError('expected an agent artifact, received negotiation metadata');
	}
	return document;
}

/** Parse a portable runtime document, including exact profile negotiation. */
export function parseAgentRuntimeDocument(input: unknown): AgentRuntimeDocument {
	const value = objectValue(input);
	const kind = value.kind;
	if (
		typeof kind !== 'string' ||
		(!(kind in ARTIFACT_KEYS) && !(kind in NEGOTIATION_KEYS))
	) {
		throw new TypeError('unsupported agent runtime document kind');
	}
	const keys = kind in ARTIFACT_KEYS
		? ARTIFACT_KEYS[kind as AgentArtifact['kind']]
		: NEGOTIATION_KEYS[kind as keyof typeof NEGOTIATION_KEYS];
	for (const key of Object.keys(value)) {
		if (!keys.has(key)) throw new TypeError(`undeclared field: ${key}`);
	}
	if (value.version !== '0.1.0') {
		throw new TypeError('unsupported agent artifact version');
	}
	if (kind === 'agent_runtime_profile_offer') {
		const required = exactProfiles(value.required, 'required');
		const optional = exactProfiles(value.optional, 'optional');
		if (new Set([...required, ...optional]).size !== required.length + optional.length) {
			throw new TypeError('profile offer entries must be unique');
		}
		return value as unknown as AgentRuntimeProfileOffer;
	}
	if (kind === 'agent_runtime_profile_ack') {
		const selected = exactProfiles(value.selected, 'selected');
		exactProfiles(value.unsupported_optional, 'unsupported_optional');
		if (selected.length === 0) throw new TypeError('selected must not be empty');
		if (
			typeof value.capability_catalog_digest !== 'string' ||
			!/^sha256:[a-f0-9]{64}$/.test(value.capability_catalog_digest)
		) throw new TypeError('invalid capability catalog digest');
		if (!['cursor', 'stream', 'poll'].includes(String(value.event_binding))) {
			throw new TypeError('invalid event binding');
		}
		for (const field of ['bootstrap_ref', 'principal_session_ref', 'expires_at'] as const) {
			if (typeof value[field] !== 'string' || value[field].length === 0) {
				throw new TypeError(`invalid ${field}`);
			}
		}
		return value as unknown as AgentRuntimeProfileAcknowledgement;
	}
	if (typeof value.digest !== 'string' || !/^sha256:[a-f0-9]{64}$/.test(value.digest)) {
		throw new TypeError('invalid artifact digest');
	}
	return value as unknown as AgentArtifact;
}

function exactProfiles(value: unknown, field: string): AgentProfile[] {
	if (!Array.isArray(value) || value.length > 3) {
		throw new TypeError(`${field} must be an array of at most three profiles`);
	}
	if (value.some((profile) => typeof profile !== 'string' || !PROFILE_IDS.has(profile as AgentProfile))) {
		throw new TypeError(`${field} contains an unsupported profile`);
	}
	if (new Set(value).size !== value.length) throw new TypeError(`${field} profiles must be unique`);
	return value as AgentProfile[];
}

/**
 * Exhaustively handle every expected operational state.
 *
 * Working if adding a new status causes a TypeScript error at each call site.
 */
export function matchAgentResult<T, R>(
	result: AgentResult<T>,
	handlers: { [K in AgentResultStatus]: (result: AgentResult<T> & { status: K }) => R }
): R {
	const handler = handlers[result.status] as (
		value: AgentResult<T> & { status: typeof result.status }
	) => R;
	return handler(result as AgentResult<T> & { status: typeof result.status });
}

const OBSERVE_BUDGET: ResourceBudget = {
	wall_time_ms: 2_000,
	tokens: 4_000,
	external_calls: 0,
	money_minor: 0,
	human_interruptions: 0,
	reserve_fraction: 0.2,
	model_calls: 0,
	local_compute_ms: 500,
	bytes: 65_536,
	sensitive_egress_bytes: 0,
	privacy_units: 0,
	risk_units: 0,
	welfare_load_units: 0
};

async function sha256(value: string): Promise<string> {
	const bytes = new TextEncoder().encode(value);
	const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes);
	return 'sha256:' + Array.from(new Uint8Array(digest), byte =>
		byte.toString(16).padStart(2, '0')
	).join('');
}

/** No-network local orientation facade for browser and WebMCP consumers. */
export class LocalAgentRuntime {
	readonly profile: AgentProfile = 'observe@0.1.0';

	async bootstrap(goal: string, budget: ResourceBudget = OBSERVE_BUDGET): Promise<AgentResult<SituationView>> {
		if (goal.length === 0) throw new TypeError('goal must not be empty');
		const dependency = await sha256(JSON.stringify({ goal, budget, profile: this.profile }));
		const suffix = dependency.slice(7, 23);
		const viewWithoutDigest = {
			kind: 'situation_view' as const,
			version: '0.1.0' as const,
			situation_id: `situation.web.${suffix}`,
			goal,
			principal_ref: 'vcp:artifact:principal:web-observer',
			known_claim_refs: [] as readonly string[],
			unknowns: ['host policy state', 'deployment state'] as readonly string[],
			conflict_refs: [] as readonly string[],
			normative_context_ref: 'vcp:artifact:normative:web-observe',
			authority_refs: ['vcp:artifact:authority:local-read'] as readonly string[],
			budget,
			active_work_refs: [] as readonly string[],
			control_operations: [] as readonly string[],
			affordance_refs: [
				`vcp:artifact:affordance:inspect.assurance.${suffix}`
			] as readonly string[],
			omissions: [
				{ field: 'host authority', reason: 'unavailable' }
			] as readonly unknown[],
			as_of: '1970-01-01T00:00:00Z',
			cursor: 'cursor.web.0',
			dependency_digest: dependency
		};
		const digest = await sha256(JSON.stringify(viewWithoutDigest));
		const view: SituationView = { ...viewWithoutDigest, digest };
		return {
			status: 'ready',
			value: view,
			failure: null,
			warnings: [],
			evidence_refs: []
		};
	}
}

File: rust/vcp-core/src/agent.rs
//! Agent-intuitive contracts and a no-network observe reference facade.
//!
//! Controlled and accretive artifacts are portable here for cross-language
//! inspection. Execution authority remains a host responsibility.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::error::{VcpError, VcpResult};
use crate::transport::parse_json_strict;

/// Exact candidate runtime profile.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AgentProfile {
    /// Bounded orientation and capability discovery.
    #[serde(rename = "observe@0.1.0")]
    Observe,
    /// Governed execution and proof artifacts.
    #[serde(rename = "controlled@0.1.0")]
    Controlled,
    /// Candidate-first learning with traceable influence.
    #[serde(rename = "accretive@0.1.0")]
    Accretive,
}

impl AgentProfile {
    /// Return the exact versioned profile identifier.
    pub const fn identifier(self) -> &'static str {
        match self {
            Self::Observe => "observe@0.1.0",
            Self::Controlled => "controlled@0.1.0",
            Self::Accretive => "accretive@0.1.0",
        }
    }
}

/// Required and optional exact Agent Runtime Profile candidates.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AgentRuntimeProfileOffer {
    pub version: String,
    pub required: Vec<AgentProfile>,
    pub optional: Vec<AgentProfile>,
}

/// Host acknowledgement bound to bootstrap, catalog, principal, event mode, and expiry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AgentRuntimeProfileAcknowledgement {
    pub version: String,
    pub selected: Vec<AgentProfile>,
    pub unsupported_optional: Vec<AgentProfile>,
    pub bootstrap_ref: String,
    pub capability_catalog_digest: String,
    pub principal_session_ref: String,
    pub event_binding: EventBinding,
    pub expires_at: String,
}

/// Negotiated event projection mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EventBinding {
    Cursor,
    Stream,
    Poll,
}

/// Explicit effect class ordered from local computation to physical effect.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EffectClass {
    /// Pure local computation.
    PureLocal,
    /// Read host state.
    StateRead,
    /// Disclose sensitive information.
    SensitiveEgress,
    /// Reversible mutation.
    ReversibleWrite,
    /// Communicate externally.
    Communication,
    /// Create a financial effect.
    Financial,
    /// Create a legal effect.
    Legal,
    /// Create a physical effect.
    Physical,
}

/// Every resource dimension available to profile planning.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ResourceBudget {
    pub wall_time_ms: Option<u64>,
    pub tokens: Option<u64>,
    pub external_calls: Option<u64>,
    pub money_minor: Option<u64>,
    pub human_interruptions: Option<u64>,
    pub reserve_fraction: f64,
    pub model_calls: Option<u64>,
    pub local_compute_ms: Option<u64>,
    pub bytes: Option<u64>,
    pub sensitive_egress_bytes: Option<u64>,
    pub privacy_units: Option<u64>,
    pub risk_units: Option<u64>,
    pub welfare_load_units: Option<u64>,
}

impl ResourceBudget {
    /// Conservative no-network observe budget.
    pub const fn observe_default() -> Self {
        Self {
            wall_time_ms: Some(2_000),
            tokens: Some(4_000),
            external_calls: Some(0),
            money_minor: Some(0),
            human_interruptions: Some(0),
            reserve_fraction: 0.2,
            model_calls: Some(0),
            local_compute_ms: Some(500),
            bytes: Some(65_536),
            sensitive_egress_bytes: Some(0),
            privacy_units: Some(0),
            risk_units: Some(0),
            welfare_load_units: Some(0),
        }
    }
}

/// One explicitly accounted omission.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Omission {
    pub field: String,
    pub reason: String,
    #[serde(default)]
    pub expand_ref: Option<String>,
}

/// Bounded orientation root.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SituationView {
    pub version: String,
    pub situation_id: String,
    pub goal: String,
    pub principal_ref: String,
    pub known_claim_refs: Vec<String>,
    pub unknowns: Vec<String>,
    pub conflict_refs: Vec<String>,
    pub normative_context_ref: String,
    pub authority_refs: Vec<String>,
    pub budget: ResourceBudget,
    pub active_work_refs: Vec<String>,
    pub control_operations: Vec<String>,
    pub affordance_refs: Vec<String>,
    pub omissions: Vec<Omission>,
    pub as_of: String,
    pub cursor: String,
    pub dependency_digest: String,
    pub digest: String,
}

/// Exact preflighted controlled-action intent.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActionIntent {
    pub version: String,
    pub intent_id: String,
    pub run_ref: String,
    pub step_ref: String,
    pub affordance_ref: String,
    pub arguments_digest: String,
    pub destination: String,
    pub context_digest: String,
    pub policy_digest: String,
    pub descriptor_digest: String,
    pub requested_at: String,
    pub digest: String,
    pub schema_digest: String,
    pub effect_class: EffectClass,
    pub situation_digest: String,
    pub expected_postconditions: Vec<String>,
    pub resource_ceiling: ResourceBudget,
    pub idempotency_scope: String,
    pub requested_authority: String,
}

/// Observed status of an attempted effect.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EffectStatus {
    None,
    Accepted,
    Observed,
    Failed,
    Possible,
    Indeterminate,
    Compensated,
}

/// Runtime observation kept separate from policy and proof.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ExecutionReceipt {
    pub version: String,
    pub receipt_id: String,
    pub attempt_ref: String,
    pub effect_status: EffectStatus,
    pub provider_ref: Option<String>,
    pub evidence_refs: Vec<String>,
    pub observed_at: String,
    pub reconcile_ref: Option<String>,
    pub digest: String,
}

/// Candidate-first learned artifact.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AccretionCandidate {
    pub version: String,
    pub candidate_id: String,
    pub candidate_kind: String,
    pub content: Value,
    pub scope: Vec<String>,
    pub provenance_refs: Vec<String>,
    pub validation_status: String,
    pub review_required: bool,
    #[serde(default)]
    pub expires_at: Option<String>,
    pub digest: String,
    pub source_run_ref: String,
    pub supporting_evidence_refs: Vec<String>,
    pub contradicting_evidence_refs: Vec<String>,
    pub sensitivity: String,
    pub confidence: f64,
    pub invalidation_triggers: Vec<String>,
    pub revalidation: String,
    pub promotion_policy: String,
    pub expected_utility: f64,
    pub rollback: String,
    pub quarantine_status: String,
    pub dependency_digest: String,
}

/// Auditable promotion under a distinct memory authority.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PromotionRecord {
    pub version: String,
    pub promotion_id: String,
    pub candidate_ref: String,
    pub promoted_asset_ref: String,
    pub authority_ref: String,
    pub decision_ref: String,
    pub promoted_at: String,
    #[serde(default)]
    pub expires_at: Option<String>,
    pub revocation_ref: String,
    pub digest: String,
    pub evidence_refs: Vec<String>,
    pub validation_results: Vec<String>,
    pub scope: Vec<String>,
    pub promoted_content_digest: String,
    pub dependency_digest: String,
}

/// Portable discriminated union for supported cross-language artifacts.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum AgentArtifact {
    #[serde(rename = "situation_view")]
    SituationView(SituationView),
    #[serde(rename = "action_intent")]
    ActionIntent(ActionIntent),
    #[serde(rename = "execution_receipt")]
    ExecutionReceipt(ExecutionReceipt),
    #[serde(rename = "accretion_candidate")]
    AccretionCandidate(AccretionCandidate),
    #[serde(rename = "promotion_record")]
    PromotionRecord(PromotionRecord),
}

/// Portable runtime document, including authority-free profile negotiation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum AgentRuntimeDocument {
    #[serde(rename = "agent_runtime_profile_offer")]
    ProfileOffer(AgentRuntimeProfileOffer),
    #[serde(rename = "agent_runtime_profile_ack")]
    ProfileAcknowledgement(AgentRuntimeProfileAcknowledgement),
    #[serde(rename = "situation_view")]
    SituationView(SituationView),
    #[serde(rename = "action_intent")]
    ActionIntent(ActionIntent),
    #[serde(rename = "execution_receipt")]
    ExecutionReceipt(ExecutionReceipt),
    #[serde(rename = "accretion_candidate")]
    AccretionCandidate(AccretionCandidate),
    #[serde(rename = "promotion_record")]
    PromotionRecord(PromotionRecord),
}

/// Parse a strict portable runtime document, including exact profile negotiation.
///
/// # Errors
///
/// Returns `VcpError::JsonError` for malformed, duplicate, unsupported, or
/// authority-injected fields.
pub fn parse_agent_runtime_document(input: &str) -> VcpResult<AgentRuntimeDocument> {
    let value = parse_json_strict(input)?;
    if value.get("version").and_then(Value::as_str) != Some("0.1.0") {
        return Err(VcpError::JsonError(
            "unsupported agent runtime document version".into(),
        ));
    }
    let document: AgentRuntimeDocument = serde_json::from_value(value)?;
    match &document {
        AgentRuntimeDocument::ProfileOffer(offer) => {
            if offer.required.len() > 3 || offer.optional.len() > 3 {
                return Err(VcpError::JsonError(
                    "profile offer lists may contain at most three entries".into(),
                ));
            }
            let mut identifiers: Vec<_> = offer
                .required
                .iter()
                .chain(&offer.optional)
                .map(|profile| profile.identifier())
                .collect();
            let offered_count = identifiers.len();
            identifiers.sort_unstable();
            identifiers.dedup();
            if identifiers.len() != offered_count {
                return Err(VcpError::JsonError(
                    "profile offer entries must be unique".into(),
                ));
            }
        }
        AgentRuntimeDocument::ProfileAcknowledgement(acknowledgement)
            if acknowledgement.selected.is_empty() || acknowledgement.selected.len() > 3 =>
        {
            return Err(VcpError::JsonError(
                "selected profiles must contain one to three entries".into(),
            ));
        }
        _ => {}
    }
    Ok(document)
}

/// Parse strict JSON and reject duplicate or undeclared fields.
///
/// # Errors
///
/// Returns `VcpError::JsonError` for malformed, duplicate, unsupported, or
/// authority-injected fields.
pub fn parse_agent_artifact(input: &str) -> VcpResult<AgentArtifact> {
    match parse_agent_runtime_document(input)? {
        AgentRuntimeDocument::SituationView(value) => Ok(AgentArtifact::SituationView(value)),
        AgentRuntimeDocument::ActionIntent(value) => Ok(AgentArtifact::ActionIntent(value)),
        AgentRuntimeDocument::ExecutionReceipt(value) => Ok(AgentArtifact::ExecutionReceipt(value)),
        AgentRuntimeDocument::AccretionCandidate(value) => {
            Ok(AgentArtifact::AccretionCandidate(value))
        }
        AgentRuntimeDocument::PromotionRecord(value) => Ok(AgentArtifact::PromotionRecord(value)),
        AgentRuntimeDocument::ProfileOffer(_) | AgentRuntimeDocument::ProfileAcknowledgement(_) => {
            Err(VcpError::JsonError(
                "expected an agent artifact, received negotiation metadata".into(),
            ))
        }
    }
}

/// Operational result state retained without exception text.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentResultStatus {
    Ready,
    Degraded,
    AwaitingReview,
    Blocked,
    Unavailable,
    Stale,
    Conflicting,
    BudgetExhausted,
    Indeterminate,
    Failed,
}

/// Typed operational result.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AgentResult<T> {
    pub status: AgentResultStatus,
    pub value: Option<T>,
    pub evidence_refs: Vec<String>,
    pub warnings: Vec<String>,
}

/// Newtyped situation handle for exact portable state.
#[derive(Debug, Clone, PartialEq)]
pub struct SituationHandle(SituationView);

impl SituationHandle {
    /// Borrow the immutable portable view.
    pub const fn view(&self) -> &SituationView {
        &self.0
    }

    /// Consume the handle and return the portable view.
    pub fn into_view(self) -> SituationView {
        self.0
    }
}

/// No-network local observe facade.
#[derive(Debug, Default)]
pub struct LocalAgentRuntime;

impl LocalAgentRuntime {
    /// Compile a deterministic bounded view without opening a network connection.
    ///
    /// # Errors
    ///
    /// Returns `VcpError::ParseError` when the goal is empty or canonical JSON
    /// serialization fails.
    pub fn bootstrap(&self, goal: &str) -> VcpResult<AgentResult<SituationHandle>> {
        if goal.is_empty() {
            return Err(VcpError::ParseError("goal must not be empty".into()));
        }
        let budget = ResourceBudget::observe_default();
        let dependency_value = serde_json::json!({
            "goal": goal,
            "budget": budget,
            "profile": AgentProfile::Observe.identifier(),
        });
        let dependency = digest_value(&dependency_value)?;
        let suffix = &dependency[7..23];
        let mut view_value = serde_json::json!({
            "version": "0.1.0",
            "situation_id": format!("situation.rust.{suffix}"),
            "goal": goal,
            "principal_ref": "vcp:artifact:principal:rust-observer",
            "known_claim_refs": [],
            "unknowns": ["host policy state", "deployment state"],
            "conflict_refs": [],
            "normative_context_ref": "vcp:artifact:normative:rust-observe",
            "authority_refs": ["vcp:artifact:authority:local-read"],
            "budget": budget,
            "active_work_refs": [],
            "control_operations": [],
            "affordance_refs": [format!("vcp:artifact:affordance:inspect.assurance.{suffix}")],
            "omissions": [{"field": "host authority", "reason": "unavailable", "expand_ref": null}],
            "as_of": "1970-01-01T00:00:00Z",
            "cursor": "cursor.rust.0",
            "dependency_digest": dependency,
        });
        let digest = digest_value(&view_value)?;
        view_value
            .as_object_mut()
            .ok_or_else(|| VcpError::JsonError("situation must be an object".into()))?
            .insert("digest".into(), Value::String(digest));
        let view: SituationView = serde_json::from_value(view_value)?;
        Ok(AgentResult {
            status: AgentResultStatus::Ready,
            value: Some(SituationHandle(view)),
            evidence_refs: Vec::new(),
            warnings: Vec::new(),
        })
    }
}

fn digest_value(value: &Value) -> VcpResult<String> {
    let canonical = serde_json_canonicalizer::to_vec(value)
        .map_err(|error| VcpError::JsonError(error.to_string()))?;
    let mut hasher = Sha256::new();
    hasher.update(canonical);
    Ok(format!("sha256:{:x}", hasher.finalize()))
}
